veracruz-server/src/server.rs

```diff
@@ -17,46 +17,167 @@ use crate::veracruz_server_sgx::veracruz_server_sgx::VeracruzServerSGX as Veracr
+                    .filter(|(_, enclave)| enclave.policy_hash == hash)
+                    .next()
+                    .map(|(id, _)| *id)
+                    .unwrap_or(0)
+            )
+        } else {
+            Ok(
+                self.enclaves.iter()
+                    .next()
+                    .map(|(id, _)| *id)
+                    .unwrap_or(0)
+            )
+        }
+    }
+}
+
+
+/// Setup an enclave with a provided policy file
+#[post("/enclave_setup")]
+async fn enclave_setup(
+    policy_json: String,
+    enclave_handler: web::Data<Arc<Mutex<EnclaveHandler>>>,
+) -> Result<String, VeracruzServerError> {
+    // parse policy
+    let (_, policy_hash) = Policy::and_hash_from_json(&policy_json)?;
+
+    // create new enclave, with policy
+    let enclave = VeracruzServerEnclave::new(&policy_json)?;
+
+    // find next id
+    let mut enclave_handler = enclave_handler.lock()?;
+    let id = enclave_handler.id_gen.next().unwrap();
+
+    // store id in list
+    enclave_handler.enclaves.insert(id, EnclaveState {
+        enclave: Box::new(enclave),
+        start_time: SystemTime::now(),
+        policy: policy_json,
+        policy_hash: policy_hash.to_owned(),
+    });
+
+    Ok(serde_json::to_string(&id)?)
+}
+
+/// Teardown an enclave
+#[post("/enclave_teardown{id:.*}")]
+async fn enclave_teardown(
+    enclave_handler: web::Data<Arc<Mutex<EnclaveHandler>>>,
+    req: HttpRequest,
+) -> Result<String, VeracruzServerError> {
+    // does enclave exist?
+    let mut enclave_handler = enclave_handler.lock()?;
+    let id = enclave_handler.best_id(req.query_string())?;
+    match enclave_handler.enclaves.remove(&id) {
+        Some(mut enclave) => {
+            // well it doesn't anymore, let close take care of the rest
+            enclave.enclave.close()?;
+            Ok("".to_owned())
+        }
+        None => {
+            Err(VeracruzServerError::InvalidEnclaveError(id))?
+        }
+    }
+}
+
+/// Query a list of running enclaves
+///
+/// Currently this can only be one or zero enclaves
+///
+#[get("/enclave_list")]
+async fn enclave_list(
+    enclave_handler: web::Data<Arc<Mutex<EnclaveHandler>>>,
+) -> Result<String, VeracruzServerError> {
+    let mut enclave_list = Vec::<serde_json::Value>::new();
     
-    let mut enclave_handler_locked = enclave_handler.lock()?;
+    let enclave_handler = enclave_handler.lock()?;
+    for (id, enclave) in enclave_handler.enclaves.iter() {
+        // Apparently SystemTime can error if time has moved backwards,
+        // if this happens (clock change?) we just show an uptime of zero
+        let uptime = enclave.start_time.elapsed()
+            .unwrap_or_else(|_| Duration::from_secs(0));
 
-    let enclave = enclave_handler_locked.as_mut().ok_or(VeracruzServerError::UninitializedEnclaveError)?;
+        // note we can't return a web::Json object here because Actix and Veracruz
+        // are actually using two incompatible versions of serde at the moment
```

Comment:
Oof.  Is this an easy fix?

Reply:
This may be much easier to fix now with the workspaces.